File: src/ark/schema_hub.py
"""
ARK Schema Hub — Community Schema Registry
v0.4.0: 社区驱动的Schema生态中心

设计哲学:
  1. 零摩擦贡献 — 一个Python文件即一个Schema
  2. 离线/在线双模 — 本地缓存 + 远程Hub
  3. 版本化 — 每个Schema带语义化版本
  4. 可发现 — 按类别、标签、评分搜索
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Optional, Type

from pydantic import BaseModel, Field


@dataclass
class SchemaMeta:
    """Schema元数据"""
    name: str
    version: str = "1.0.0"
    author: str = "community"
    category: str = "general"
    tags: List[str] = field(default_factory=list)
    description: str = ""
    source: str = "local"  # local | remote
    downloads: int = 0
    rating: float = 0.0

    def to_dict(self) -> Dict[str, Any]:
        return {
            "name": self.name,
            "version": self.version,
            "author": self.author,
            "category": self.category,
            "tags": self.tags,
            "description": self.description,
            "source": self.source,
            "downloads": self.downloads,
            "rating": self.rating,
        }


# ─── 内置类别 ───
CATEGORIES = [
    "payment",     # 支付
    "email",       # 邮件
    "github",      # GitHub
    "database",    # 数据库
    "http",        # HTTP/API
    "file",        # 文件操作
    "messaging",   # 消息
    "project",     # 项目管理
    "ai",          # AI/ML
    "security",    # 安全
    "general",     # 通用
]


class SchemaHub:
    """
    Community Schema Hub — 社区驱动的Schema注册与发现中心

    用法:
        hub = SchemaHub()
        hub.register(my_schema)  # 注册本地Schema
        hub.import_dir("./schemas")  # 批量加载
        results = hub.search(category="payment")  # 按类别搜索
        results = hub.search(tags=["stripe", "refund"])  # 按标签搜索
        hub.export_json("schemas.json")  # 导出供他人使用
    """

    def __init__(self, schemas_dir: Optional[str] = None):
        self._schemas: Dict[str, Type[BaseModel]] = {}
        self._meta: Dict[str, SchemaMeta] = {}
        self._schemas_dir = Path(schemas_dir) if schemas_dir else None
        self._register_builtins()

    # ═══════════════════════════════════════════════
    # Registration
    # ═══════════════════════════════════════════════

    def register(
        self,
        name: str,
        schema_cls: Type[BaseModel],
        meta: Optional[SchemaMeta] = None,
    ) -> None:
        """注册一个Schema到Hub"""
        self._schemas[name] = schema_cls
        if meta is None:
            meta = SchemaMeta(name=name)
        self._meta[name] = meta

    def _register_builtins(self) -> None:
        """注册13个内置Schema"""

        # === 支付 ===
        class StripeCharge(BaseModel):
            amount: float = Field(gt=0, description="Charge amount in cents")
            currency: str = Field(default="usd", pattern=r"^[a-z]{3}$")
            description: Optional[str] = Field(default=None)
            customer: Optional[str] = Field(default=None)

        class StripeRefund(BaseModel):
            charge_id: str = Field(min_length=5)
            amount: Optional[float] = Field(default=None, gt=0)
            reason: Optional[str] = Field(default=None)

        # === 邮件 ===
        class SendEmail(BaseModel):
            to: str = Field(pattern=r"^[^@\s]+@[^@\s]+\.[^@\s]+$")
            subject: str = Field(min_length=1, max_length=998)
            body: str = Field(min_length=1)
            cc: Optional[List[str]] = Field(default=None)

        class SendBulkEmail(BaseModel):
            to: List[str] = Field(min_length=1)
            subject: str = Field(min_length=1)
            body: str = Field(min_length=1)

        # === GitHub ===
        class GitHubCreateIssue(BaseModel):
            owner: str = Field(min_length=1)
            repo: str = Field(min_length=1)
            title: str = Field(min_length=1, max_length=256)
            body: Optional[str] = Field(default="")
            labels: Optional[List[str]] = Field(default=None)

        class GitHubCreatePR(BaseModel):
            owner: str
            repo: str
            title: str = Field(min_length=1)
            head: str = Field(min_length=1)
            base: str = Field(default="main")
            body: Optional[str] = Field(default="")

        # === 数据库 ===
        class SQLQuery(BaseModel):
            query: str = Field(min_length=1)
            params: Optional[Dict] = Field(default=None)

        class SQLInsert(BaseModel):
            table: str = Field(min_length=1, pattern=r"^[a-zA-Z_][a-zA-Z0-9_]*$")
            values: Dict = Field(min_length=1)

        # === HTTP ===
        class HTTPRequest(BaseModel):
            url: str = Field(pattern=r"^https?://")
            method: str = Field(default="GET", pattern=r"^(GET|POST|PUT|DELETE|PATCH)$")
            headers: Optional[Dict] = Field(default=None)
            body: Optional[Dict] = Field(default=None)

        # === 文件 ===
        class FileRead(BaseModel):
            path: str = Field(min_length=1)
            encoding: str = Field(default="utf-8")

        class FileWrite(BaseModel):
            path: str = Field(min_length=1)
            content: str
            mode: str = Field(default="w", pattern=r"^(w|a|x)$")

        # === Slack ===
        class SlackMessage(BaseModel):
            channel: str = Field(min_length=1)
            text: str = Field(min_length=1)
            thread_ts: Optional[str] = Field(default=None)

        # === Jira ===
        class JiraCreateTicket(BaseModel):
            project: str = Field(min_length=1, max_length=10)
            summary: str = Field(min_length=1)
            description: Optional[str] = Field(default="")
            issue_type: str = Field(default="Bug", pattern=r"^(Bug|Task|Story|Epic)$")
            priority: str = Field(default="Medium", pattern=r"^(Highest|High|Medium|Low|Lowest)$")

        # 注册内置Schema + 元数据
        builtins: Dict[str, tuple[Type[BaseModel], SchemaMeta]] = {
            "stripe.charge": (StripeCharge, SchemaMeta(
                name="stripe.charge", version="1.0.0", author="ark-core",
                category="payment", tags=["stripe", "charge", "payment"],
                description="Stripe charge request schema"
            )),
            "stripe.refund": (StripeRefund, SchemaMeta(
                name="stripe.refund", version="1.0.0", author="ark-core",
                category="payment", tags=["stripe", "refund", "payment"],
                description="Stripe refund request schema"
            )),
            "email.send": (SendEmail, SchemaMeta(
                name="email.send", version="1.0.0", author="ark-core",
                category="email", tags=["email", "send"],
                description="Send single email schema"
            )),
            "email.send_bulk": (SendBulkEmail, SchemaMeta(
                name="email.send_bulk", version="1.0.0", author="ark-core",
                category="email", tags=["email", "bulk", "send"],
                description="Send bulk emails schema"
            )),
            "github.create_issue": (GitHubCreateIssue, SchemaMeta(
                name="github.create_issue", version="1.0.0", author="ark-core",
                category="github", tags=["github", "issue"],
                description="GitHub create issue schema"
            )),
            "github.create_pr": (GitHubCreatePR, SchemaMeta(
                name="github.create_pr", version="1.0.0", author="ark-core",
                category="github", tags=["github", "pr", "pull request"],
                description="GitHub create pull request schema"
            )),
            "db.query": (SQLQuery, SchemaMeta(
                name="db.query", version="1.0.0", author="ark-core",
                category="database", tags=["sql", "query", "database"],
                description="SQL query schema with parameterized params"
            )),
            "db.insert": (SQLInsert, SchemaMeta(
                name="db.insert", version="1.0.0", author="ark-core",
                category="database", tags=["sql", "insert", "database"],
                description="SQL insert schema with table validation"
            )),
            "http.request": (HTTPRequest, SchemaMeta(
                name="http.request", version="1.0.0", author="ark-core",
                category="http", tags=["http", "api", "request"],
                description="HTTP API request schema"
            )),
            "file.read": (FileRead, SchemaMeta(
                name="file.read", version="1.0.0", author="ark-core",
                category="file", tags=["file", "read"],
                description="File read schema"
            )),
            "file.write": (FileWrite, SchemaMeta(
                name="file.write", version="1.0.0", author="ark-core",
                category="file", tags=["file", "write"],
                description="File write schema"
            )),
            "slack.message": (SlackMessage, SchemaMeta(
                name="slack.message", version="1.0.0", author="ark-core",
                category="messaging", tags=["slack", "message", "chat"],
                description="Slack message send schema"
            )),
            "jira.create_ticket": (JiraCreateTicket, SchemaMeta(
                name="jira.create_ticket", version="1.0.0", author="ark-core",
                category="project", tags=["jira", "ticket", "project"],
                description="Jira create ticket schema"
            )),
        }

        for name, (cls, meta) in builtins.items():
            self.register(name, cls, meta)

    # ═══════════════════════════════════════════════
    # Discovery
    # ═══════════════════════════════════════════════

    def get(self, name: str) -> Optional[Type[BaseModel]]:
        """获取单个Schema"""
        return self._schemas.get(name)

    def get_meta(self, name: str) -> Optional[SchemaMeta]:
        """获取Schema元数据"""
        return self._meta.get(name)

    @property
    def available(self) -> List[str]:
        """所有已注册Schema名称"""
        return sorted(self._schemas.keys())

    @property
    def categories(self) -> List[str]:
        """所有类别"""
        seen = set()
        for m in self._meta.values():
            seen.add(m.category)
        return sorted(seen)

    def search(
        self,
        query: Optional[str] = None,
        category: Optional[str] = None,
        tags: Optional[List[str]] = None,
        author: Optional[str] = None,
    ) -> List[SchemaMeta]:
        """
        多维度搜索Schema

        Args:
            query: 名称/描述中包含的文字
            category: 按类别过滤
            tags: 按标签过滤 (AND逻辑)
            author: 按作者过滤

        Returns:
            匹配的SchemaMeta列表
        """
        results = []
        for name, meta in self._meta.items():
            # query 匹配
            if query and query.lower() not in name.lower() and query.lower() not in meta.description.lower():
                continue
            # category 过滤
            if category and meta.category != category:
                continue
            # tags 过滤 (AND)
            if tags and not all(t in meta.tags for t in tags):
                continue
            # author 过滤
            if author and meta.author != author:
                continue
            results.append(meta)
        return results

    def list_by_category(self) -> Dict[str, List[SchemaMeta]]:
        """按类别分组列出所有Schema"""
        result: Dict[str, List[SchemaMeta]] = {}
        for meta in self._meta.values():
            result.setdefault(meta.category, []).append(meta)
        return result

    # ═══════════════════════════════════════════════
    # Import / Export
    # ═══════════════════════════════════════════════

    def import_dir(self, dir_path: str) -> int:
        """
        从目录批量加载Schema JSON文件

        Schema JSON格式:
        {
            "name": "tool_name",
            "version": "1.0.0",
            "author": "community_user",
            "category": "ai",
            "tags": ["openai", "chat"],
            "description": "...",
            "fields": {
                "field_name": {"type": "str", "required": true, "description": "..."},
                ...
            }
        }

        返回加载数量
        """
        import json as json_module
        from pydantic import create_model

        d = Path(dir_path)
        if not d.is_dir():
            return 0

        count = 0
        for f in d.glob("*.json"):
            try:
                data = json_module.loads(f.read_text())
                fields_data = data.get("fields", {})
                fields = {}
                for fname, finfo in fields_data.items():
                    t = finfo.get("type", "str")
                    required = finfo.get("required", True)
                    desc = finfo.get("description", "")
                    if t == "str":
                        fields[fname] = (str, Field(default=... if required else None, description=desc))
                    elif t == "int":
                        fields[fname] = (int, Field(default=... if required else None, description=desc))
                    elif t == "float":
                        fields[fname] = (float, Field(default=... if required else None, description=desc))
                    elif t == "bool":
                        fields[fname] = (bool, Field(default=... if required else None, description=desc))
                    elif t.startswith("list["):
                        fields[fname] = (List, Field(default=... if required else None, description=desc))
                    elif t.startswith("dict"):
                        fields[fname] = (Dict, Field(default=... if required else None, description=desc))
                    else:
                        fields[fname] = (Optional[str], Field(default=None, description=desc))

                if fields:
                    model = create_model(data["name"].replace(".", "_"), **fields)
                    meta = SchemaMeta(
                        name=data["name"],
                        version=data.get("version", "1.0.0"),
                        author=data.get("author", "community"),
                        category=data.get("category", "general"),
                        tags=data.get("tags", []),
                        description=data.get("description", ""),
                        source="local",
                    )
                    self.register(data["name"], model, meta)
                    count += 1
            except Exception:
                continue

        return count

    def export_json(self, path: str) -> None:
        """导出所有Schema为JSON (供社区导入)"""
        result = []
        for name, schema_cls in self._schemas.items():
            meta = self._meta.get(name)
            entry = {
                "name": name,
                "version": meta.version if meta else "1.0.0",
                "author": meta.author if meta else "community",
                "category": meta.category if meta else "general",
                "tags": meta.tags if meta else [],
                "description": meta.description if meta else "",
                "json_schema": schema_cls.model_json_schema(),
            }
            result.append(entry)
        with open(path, "w") as f:
            json.dump(result, f, indent=2, ensure_ascii=False)

    def export_meta_json(self, path: str) -> None:
        """导出所有元数据为JSON (轻量版，供搜索索引)"""
        result = [m.to_dict() for m in self._meta.values()]
        with open(path, "w") as f:
            json.dump(result, f, indent=2, ensure_ascii=False)

    # ═══════════════════════════════════════════════
    # Validation
    # ═══════════════════════════════════════════════

    def validate(self, name: str, data: Any) -> Optional[BaseModel]:
        """使用Hub中的Schema验证数据"""
        schema = self.get(name)
        if schema is None:
            return None
        try:
            return schema(**data) if isinstance(data, dict) else schema.model_validate(data)
        except Exception:
            return None

    # ═══════════════════════════════════════════════
    # Stats
    # ═══════════════════════════════════════════════

    @property
    def stats(self) -> Dict[str, Any]:
        """Hub统计信息"""
        cats = self.list_by_category()
        return {
            "total_schemas": len(self._schemas),
            "categories": len(cats),
            "by_category": {k: len(v) for k, v in cats.items()},
            "total_tags": len(set(t for m in self._meta.values() for t in m.tags)),
            "authors": len(set(m.author for m in self._meta.values())),
        }

    def __repr__(self) -> str:
        return f"SchemaHub(schemas={len(self._schemas)}, categories={len(self.categories)})"


# ─── 全局单例 ───
_hub: Optional[SchemaHub] = None


def get_schema_hub(schemas_dir: Optional[str] = None) -> SchemaHub:
    """获取全局SchemaHub单例"""
    global _hub
    if _hub is None:
        _hub = SchemaHub(schemas_dir=schemas_dir)
    return _hub

File: sdk/go/ark/schema_hub.go
package ark

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
)

// SchemaMeta holds metadata for a registered schema.
type SchemaMeta struct {
	Name        string   `json:"name"`
	Version     string   `json:"version"`
	Author      string   `json:"author"`
	Category    string   `json:"category"`
	Tags        []string `json:"tags"`
	Description string   `json:"description"`
	Source      string   `json:"source"` // "local" | "remote"
	Downloads   int      `json:"downloads"`
	Rating      float64  `json:"rating"`
}

// SchemaHub is a community-driven schema registry and discovery center.
// It mirrors the Python SchemaHub with 13 built-in business schemas.
type SchemaHub struct {
	schemas map[string]ValidationSchema
	meta    map[string]SchemaMeta
	mu      sync.RWMutex
}

// Predefined categories matching the Python SDK.
var Categories = []string{
	"payment",
	"email",
	"github",
	"database",
	"http",
	"file",
	"messaging",
	"project",
	"ai",
	"security",
	"general",
}

// NewSchemaHub creates a SchemaHub with 13 built-in schemas.
func NewSchemaHub() *SchemaHub {
	h := &SchemaHub{
		schemas: make(map[string]ValidationSchema),
		meta:    make(map[string]SchemaMeta),
	}
	h.registerBuiltins()
	return h
}

// Register adds a schema with metadata to the hub.
func (h *SchemaHub) Register(name string, schema ValidationSchema, meta SchemaMeta) {
	meta.Name = name
	if meta.Source == "" {
		meta.Source = "local"
	}
	if meta.Version == "" {
		meta.Version = "1.0.0"
	}
	if meta.Category == "" {
		meta.Category = "general"
	}
	if meta.Author == "" {
		meta.Author = "community"
	}

	h.mu.Lock()
	defer h.mu.Unlock()
	h.schemas[name] = schema
	h.meta[name] = meta
}

// Get returns a registered schema by name.
func (h *SchemaHub) Get(name string) (ValidationSchema, bool) {
	h.mu.RLock()
	defer h.mu.RUnlock()
	s, ok := h.schemas[name]
	return s, ok
}

// GetMeta returns metadata for a registered schema.
func (h *SchemaHub) GetMeta(name string) (SchemaMeta, bool) {
	h.mu.RLock()
	defer h.mu.RUnlock()
	m, ok := h.meta[name]
	return m, ok
}

// Available returns all registered schema names sorted.
func (h *SchemaHub) Available() []string {
	h.mu.RLock()
	defer h.mu.RUnlock()
	names := make([]string, 0, len(h.schemas))
	for n := range h.schemas {
		names = append(names, n)
	}
	sort.Strings(names)
	return names
}

// Categories returns all unique categories across registered schemas.
func (h *SchemaHub) Categories() []string {
	h.mu.RLock()
	defer h.mu.RUnlock()
	seen := make(map[string]bool)
	for _, m := range h.meta {
		seen[m.Category] = true
	}
	cats := make([]string, 0, len(seen))
	for c := range seen {
		cats = append(cats, c)
	}
	sort.Strings(cats)
	return cats
}

// Search finds schemas by query, category, tags, or author.
// All filters are AND logic.
func (h *SchemaHub) Search(query, category string, tags []string, author string) []SchemaMeta {
	h.mu.RLock()
	defer h.mu.RUnlock()

	var results []SchemaMeta
	for name, m := range h.meta {
		// Query filter: match against name or description
		if query != "" {
			q := strings.ToLower(query)
			if !strings.Contains(strings.ToLower(name), q) &&
				!strings.Contains(strings.ToLower(m.Description), q) {
				continue
			}
		}
		// Category filter
		if category != "" && m.Category != category {
			continue
		}
		// Tags filter (AND)
		if len(tags) > 0 {
			hasAll := true
			for _, t := range tags {
				found := false
				for _, mt := range m.Tags {
					if mt == t {
						found = true
						break
					}
				}
				if !found {
					hasAll = false
					break
				}
			}
			if !hasAll {
				continue
			}
		}
		// Author filter
		if author != "" && m.Author != author {
			continue
		}
		results = append(results, m)
	}
	return results
}

// ListByCategory groups schemas by category.
func (h *SchemaHub) ListByCategory() map[string][]SchemaMeta {
	h.mu.RLock()
	defer h.mu.RUnlock()

	result := make(map[string][]SchemaMeta)
	for _, m := range h.meta {
		result[m.Category] = append(result[m.Category], m)
	}
	return result
}

// Validate checks data against a registered schema.
func (h *SchemaHub) Validate(name string, data map[string]interface{}) (ValidationResult, error) {
	schema, ok := h.Get(name)
	if !ok {
		return ValidationResult{Passed: false}, fmt.Errorf("schema not found: %s", name)
	}
	v := NewOutputValidator()
	v.RegisterSchema(name, schema)
	return v.Validate(name, data), nil
}

// Stats returns hub-wide statistics.
func (h *SchemaHub) Stats() map[string]interface{} {
	h.mu.RLock()
	defer h.mu.RUnlock()

	cats := h.ListByCategory()
	byCategory := make(map[string]int)
	for k, v := range cats {
		byCategory[k] = len(v)
	}

	allTags := make(map[string]bool)
	authors := make(map[string]bool)
	for _, m := range h.meta {
		for _, t := range m.Tags {
			allTags[t] = true
		}
		authors[m.Author] = true
	}

	return map[string]interface{}{
		"total_schemas": len(h.schemas),
		"categories":    len(cats),
		"by_category":   byCategory,
		"total_tags":    len(allTags),
		"authors":       len(authors),
	}
}

// ─── Import / Export ───

// schemaFile represents a JSON schema definition file.
type schemaFile struct {
	Name        string               `json:"name"`
	Version     string               `json:"version"`
	Author      string               `json:"author"`
	Category    string               `json:"category"`
	Tags        []string             `json:"tags"`
	Description string               `json:"description"`
	Fields      []schemaFileField    `json:"fields"`
}

type schemaFileField struct {
	Name        string `json:"name"`
	Type        string `json:"type"`
	Required    bool   `json:"required"`
	Min         *float64 `json:"min,omitempty"`
	Max         *float64 `json:"max,omitempty"`
	Pattern     string `json:"pattern,omitempty"`
	MinLen      *int    `json:"min_len,omitempty"`
	MaxLen      *int    `json:"max_len,omitempty"`
	Description string `json:"description"`
}

// ImportDir loads all .json schema files from a directory into the hub.
// Returns the count of successfully loaded schemas.
func (h *SchemaHub) ImportDir(dir string) int {
	entries, err := filepath.Glob(filepath.Join(dir, "*.json"))
	if err != nil {
		return 0
	}

	count := 0
	for _, entry := range entries {
		data, err := os.ReadFile(entry)
		if err != nil {
			continue
		}
		var sf schemaFile
		if err := json.Unmarshal(data, &sf); err != nil {
			continue
		}
		if sf.Name == "" || len(sf.Fields) == 0 {
			continue
		}

		// Build FieldRules from file fields
		fields := make([]FieldRule, len(sf.Fields))
		for i, f := range sf.Fields {
			fields[i] = FieldRule{
				Name:     f.Name,
				Type:     f.Type,
				Required: f.Required,
				Min:      f.Min,
				Max:      f.Max,
				Pattern:  f.Pattern,
				MinLen:   f.MinLen,
				MaxLen:   f.MaxLen,
			}
		}

		meta := SchemaMeta{
			Name:        sf.Name,
			Version:     orDefault(sf.Version, "1.0.0"),
			Author:      orDefault(sf.Author, "community"),
			Category:    orDefault(sf.Category, "general"),
			Tags:        sf.Tags,
			Description: sf.Description,
			Source:      "local",
		}

		h.Register(sf.Name, ValidationSchema{Fields: fields}, meta)
		count++
	}
	return count
}

// exportEntry represents a schema for JSON export.
type exportEntry struct {
	Name        string   `json:"name"`
	Version     string   `json:"version"`
	Author      string   `json:"author"`
	Category    string   `json:"category"`
	Tags        []string `json:"tags"`
	Description string   `json:"description"`
	Fields      []exportField `json:"fields"`
}

type exportField struct {
	Name        string  `json:"name"`
	Type        string  `json:"type"`
	Required    bool    `json:"required"`
	Min         *float64 `json:"min,omitempty"`
	Max         *float64 `json:"max,omitempty"`
	Pattern     string  `json:"pattern,omitempty"`
	MinLen      *int     `json:"min_len,omitempty"`
	MaxLen      *int     `json:"max_len,omitempty"`
}

// ExportJSON writes all schemas to a JSON file (for community import).
func (h *SchemaHub) ExportJSON(path string) error {
	h.mu.RLock()
	defer h.mu.RUnlock()

	var entries []exportEntry
	for name, schema := range h.schemas {
		m := h.meta[name]
		fields := make([]exportField, len(schema.Fields))
		for i, f := range schema.Fields {
			fields[i] = exportField{
				Name:     f.Name,
				Type:     f.Type,
				Required: f.Required,
				Min:      f.Min,
				Max:      f.Max,
				Pattern:  f.Pattern,
				MinLen:   f.MinLen,
				MaxLen:   f.MaxLen,
			}
		}
		entries = append(entries, exportEntry{
			Name:        name,
			Version:     m.Version,
			Author:      m.Author,
			Category:    m.Category,
			Tags:        m.Tags,
			Description: m.Description,
			Fields:      fields,
		})
	}

	data, err := json.MarshalIndent(entries, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0644)
}

// ExportMetaJSON writes schema metadata as JSON (lightweight search index).
func (h *SchemaHub) ExportMetaJSON(path string) error {
	h.mu.RLock()
	defer h.mu.RUnlock()

	var metas []SchemaMeta
	for _, m := range h.meta {
		metas = append(metas, m)
	}
	sort.Slice(metas, func(i, j int) bool { return metas[i].Name < metas[j].Name })

	data, err := json.MarshalIndent(metas, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0644)
}

// ─── 13 Built-in Schemas ───

func (h *SchemaHub) registerBuiltins() {
	min1 := 1
	min5 := 5
	max256 := 256
	max998 := 998
	usdMin := 0.01

	builtins := []struct {
		name   string
		fields []FieldRule
		meta   SchemaMeta
	}{
		// ── Payment ──
		{
			name: "stripe.charge",
			fields: []FieldRule{
				{Name: "amount", Type: "number", Required: true, Min: &usdMin},
				{Name: "currency", Type: "string", Required: true, Pattern: `^[a-z]{3}$`},
				{Name: "description", Type: "string"},
				{Name: "customer", Type: "string"},
			},
			meta: SchemaMeta{Version: "1.0.0", Author: "ark-core", Category: "payment",
				Tags: []string{"stripe", "charge", "payment"}, Description: "Stripe charge request schema"},
		},
		{
			name: "stripe.refund",
			fields: []FieldRule{
				{Name: "charge_id", Type: "string", Required: true, MinLen: &min5},
				{Name: "amount", Type: "number", Min: &usdMin},
				{Name: "reason", Type: "string"},
			},
			meta: SchemaMeta{Version: "1.0.0", Author: "ark-core", Category: "payment",
				Tags: []string{"stripe", "refund", "payment"}, Description: "Stripe refund request schema"},
		},
		// ── Email ──
		{
			name: "email.send",
			fields: []FieldRule{
				{Name: "to", Type: "string", Required: true, Pattern: `^[^@\s]+@[^@\s]+\.[^@\s]+$`},
				{Name: "subject", Type: "string", Required: true, MinLen: &min1, MaxLen: &max998},
				{Name: "body", Type: "string", Required: true, MinLen: &min1},
				{Name: "cc", Type: "array"},
			},
			meta: SchemaMeta{Version: "1.0.0", Author: "ark-core", Category: "email",
				Tags: []string{"email", "send"}, Description: "Send single email schema"},
		},
		{
			name: "email.send_bulk",
			fields: []FieldRule{
				{Name: "to", Type: "array", Required: true},
				{Name: "subject", Type: "string", Required: true, MinLen: &min1},
				{Name: "body", Type: "string", Required: true, MinLen: &min1},
			},
			meta: SchemaMeta{Version: "1.0.0", Author: "ark-core", Category: "email",
				Tags: []string{"email", "bulk", "send"}, Description: "Send bulk emails schema"},
		},
		// ── GitHub ──
		{
			name: "github.create_issue",
			fields: []FieldRule{
				{Name: "owner", Type: "string", Required: true, MinLen: &min1},
				{Name: "repo", Type: "string", Required: true, MinLen: &min1},
				{Name: "title", Type: "string", Required: true, MinLen: &min1, MaxLen: &max256},
				{Name: "body", Type: "string"},
				{Name: "labels", Type: "array"},
			},
			meta: SchemaMeta{Version: "1.0.0", Author: "ark-core", Category: "github",
				Tags: []string{"github", "issue"}, Description: "GitHub create issue schema"},
		},
		{
			name: "github.create_pr",
			fields: []FieldRule{
				{Name: "owner", Type: "string", Required: true},
				{Name: "repo", Type: "string", Required: true},
				{Name: "title", Type: "string", Required: true, MinLen: &min1},
				{Name: "head", Type: "string", Required: true, MinLen: &min1},
				{Name: "base", Type: "string"},
				{Name: "body", Type: "string"},
			},
			meta: SchemaMeta{Version: "1.0.0", Author: "ark-core", Category: "github",
				Tags: []string{"github", "pr", "pull request"}, Description: "GitHub create pull request schema"},
		},
		// ── Database ──
		{
			name: "db.query",
			fields: []FieldRule{
				{Name: "query", Type: "string", Required: true, MinLen: &min1},
				{Name: "params", Type: "object"},
			},
			meta: SchemaMeta{Version: "1.0.0", Author: "ark-core", Category: "database",
				Tags: []string{"sql", "query", "database"}, Description: "SQL query schema with parameterized params"},
		},
		{
			name: "db.insert",
			fields: []FieldRule{
				{Name: "table", Type: "string", Required: true, MinLen: &min1, Pattern: `^[a-zA-Z_][a-zA-Z0-9_]*$`},
				{Name: "values", Type: "object", Required: true},
			},
			meta: SchemaMeta{Version: "1.0.0", Author: "ark-core", Category: "database",
				Tags: []string{"sql", "insert", "database"}, Description: "SQL insert schema with table validation"},
		},
		// ── HTTP ──
		{
			name: "http.request",
			fields: []FieldRule{
				{Name: "url", Type: "string", Required: true, Pattern: `^https?://`},
				{Name: "method", Type: "string", Required: false, Pattern: `^(GET|POST|PUT|DELETE|PATCH)$`},
				{Name: "headers", Type: "object"},
				{Name: "body", Type: "object"},
			},
			meta: SchemaMeta{Version: "1.0.0", Author: "ark-core", Category: "http",
				Tags: []string{"http", "api", "request"}, Description: "HTTP API request schema"},
		},
		// ── File ──
		{
			name: "file.read",
			fields: []FieldRule{
				{Name: "path", Type: "string", Required: true, MinLen: &min1},
				{Name: "encoding", Type: "string"},
			},
			meta: SchemaMeta{Version: "1.0.0", Author: "ark-core", Category: "file",
				Tags: []string{"file", "read"}, Description: "File read schema"},
		},
		{
			name: "file.write",
			fields: []FieldRule{
				{Name: "path", Type: "string", Required: true, MinLen: &min1},
				{Name: "content", Type: "string", Required: true},
				{Name: "mode", Type: "string", Pattern: `^(w|a|x)$`},
			},
			meta: SchemaMeta{Version: "1.0.0", Author: "ark-core", Category: "file",
				Tags: []string{"file", "write"}, Description: "File write schema"},
		},
		// ── Messaging ──
		{
			name: "slack.message",
			fields: []FieldRule{
				{Name: "channel", Type: "string", Required: true, MinLen: &min1},
				{Name: "text", Type: "string", Required: true, MinLen: &min1},
				{Name: "thread_ts", Type: "string"},
			},
			meta: SchemaMeta{Version: "1.0.0", Author: "ark-core", Category: "messaging",
				Tags: []string{"slack", "message", "chat"}, Description: "Slack message send schema"},
		},
		// ── Project ──
		{
			name: "jira.create_ticket",
			fields: []FieldRule{
				{Name: "project", Type: "string", Required: true, MinLen: &min1, MaxLen: func() *int { v := 10; return &v }()},
				{Name: "summary", Type: "string", Required: true, MinLen: &min1},
				{Name: "description", Type: "string"},
				{Name: "issue_type", Type: "string", Pattern: `^(Bug|Task|Story|Epic)$`},
				{Name: "priority", Type: "string", Pattern: `^(Highest|High|Medium|Low|Lowest)$`},
			},
			meta: SchemaMeta{Version: "1.0.0", Author: "ark-core", Category: "project",
				Tags: []string{"jira", "ticket", "project"}, Description: "Jira create ticket schema"},
		},
	}

	for _, b := range builtins {
		h.Register(b.name, ValidationSchema{Fields: b.fields}, b.meta)
	}
}

// ─── Global Singleton ───

var globalHub *SchemaHub
var hubOnce sync.Once

// GetSchemaHub returns the global SchemaHub singleton.
func GetSchemaHub() *SchemaHub {
	hubOnce.Do(func() {
		globalHub = NewSchemaHub()
	})
	return globalHub
}

// ResetSchemaHub resets the global singleton (for testing).
func ResetSchemaHub() {
	globalHub = nil
	hubOnce = sync.Once{}
}

func orDefault(s, def string) string {
	if s == "" {
		return def
	}
	return s
}
